src/ng/directive/ngModel.js

```diff
@@ -878,6 +878,141 @@ NgModelController.prototype = {
+   * - the `ng-empty` or `ng-not-empty` class is set on the element
+   * - if the $viewValue has changed
+   *   - {@link ngModel.NgModelController#$render $render} is called on the control
+   *   - the {@link ngModel.NgModelController#$validators $validators} are run and
+   *   the validation status is set.
+   *
+   * This method is called by ngModel internally when the bound scope value changes.
+   * Application developers usually do not have to call this function themselves.
+   *
+   * This function can be used when the $viewValue or the rendered DOM value of the control should
+   * be updated after a user input.
+   *
+   * For example, consider a text input with an autocomplete list (for fruit, where the items are
+   * objects with a name and an id.
+   * A user enters `ap` and then selects `Apricot` from the list.
+   * Based on this, the autocomplete widget will call $setViewValue({name: 'Apricot', id: 443}),
+   * but the rendered value will still be `ap`.
+   * The widget can then call ctrl.$processModelValue() to run the model -> view
+   * pipeline again, which includes a formatter that converts the object into the string `Apricot`
+   * which is set to the $viewValue and rendered in the DOM.
+   *
+   * @example
+   * <example module="inputExample" name="ng-model-process">
+     <file name="app.js">
+      angular.module('inputExample', [])
+        .directive('processModel', function() {
+          return {
+            require: 'ngModel',
+            link: function(scope, element, attrs, ngModel) {
+
+              ngModel.$formatters.push(function(value) {
+                if (angular.isObject(value) && value.name) {
+                  return value.name;
+                }
+
+                return value;
+              });
+
+              ngModel.$parsers.push(function(value) {
+                if (angular.isString(value)) {
+                  return scope.items.find(function(item) {
+                    return item.name === value;
+                  }) || value;
+                }
+
+                return value;
+              });
+
+              scope.items = [
+                {name: 'Apricot', id: 443},
+                {name: 'Clementine', id: 972},
+                {name: 'Durian', id: 169},
+                {name: 'Fig', id: 298},
+                {name: 'Jackfruit', id: 982},
+                {name: 'Kiwi', id: 151},
+                {name: 'Strawberry', id: 863}
+              ];
+
+              scope.select = function(item) {
+                ngModel.$setViewValue(item);
+                ngModel.$processModelValue();
+              };
+            }
+          };
+        });
+     </file>
+     <file name="index.html">
+       <div style="display: flex;">
+         <div margin-right: 30px;">
+           Search Fruit:
+           <input ng-model="val" process-model />
+
+           <ul>
+             <li ng-repeat="item in items | filter:val"><button ng-click="select(item)">{{item.name}}</li>
```

Comment:
Ehem...relying on a method set on the scope by a directive on a different element doesn't sound like a good practice :wink:
Maybe move all relevant code into the directive's template.

Reply:
I was having something like this in mind: https://plnkr.co/edit/AeOA4eMGGMONSTrLinH2?p=preview
(TBH, I find this whole usecase weird. It sounds like an unnecessary complicated way to create an autocomplete :smiley:)

Reply:
Well, that's less complex and shows better encapsulation. I'll take it. But for the record, I don't think a real world autocomplete would be successful with a hard-coded input element. At least not for me  ;)

You know the customer is always right, and apparently many devs implement autocomplete like this.

Reply:
Updated with a small tweak